core/src/main/scala/org/typelevel/sbt/TypelevelPlugin.scala

```diff
@@ -73,8 +75,8 @@ object TypelevelPlugin extends AutoPlugin {
     },
     githubWorkflowBuild := {
       WorkflowStep.Sbt(
-        List("headerCheckAll", "scalafmtCheckAll", "project /", "scalafmtSbtCheck"),
-        name = Some("Check headers and formatting"),
+        List("headerCheckAll", "scalafmtCheckAll", "scalafixAll --check", "project /", "scalafmtSbtCheck"),
+        name = Some("Check headers, migrations and formatting"),
```

Comment:
So actually after the discussion in https://github.com/typelevel/sbt-typelevel/issues/247#issuecomment-1092004966 I had an idea to do these differently.

Basically, I think the `TypelevelCiPlugin` should expose boolean settings for enabling header, fmt, scalafix, and mima checks in CI and then generate the CI steps based on those. Then, builds and/or other plugins can enable those settings (or if we are really clever, they can introspect plugins available on the build and set themselves).

This will also make it easier for an individual build to decide whether to opt-in or opt-out of any of these particular checks.

Reply:
Oh, another thing to note here, is that `scalafixAll --check` I believe needs a full compile of the project, which is why it's usually called as a later step in the build.

Reply:
Ah of course, the semanticdb doesn't exist until then anyway! I guess it ought to be a build step then.

Reply:
I had a quick look at this and this is where I got to:
```diff
diff --git a/ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala b/ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala
index 335f99a..53385fa 100644
--- a/ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala
+++ b/ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala
@@ -20,33 +20,87 @@ import sbt._
 import org.typelevel.sbt.gha.GenerativePlugin
 import org.typelevel.sbt.gha.GitHubActionsPlugin
 import org.typelevel.sbt.gha.GenerativePlugin.autoImport._
+import org.typelevel.sbt.TypelevelScalafixPlugin.autoImport._
 import com.typesafe.tools.mima.plugin.MimaPlugin
 import scala.language.experimental.macros
 
 object TypelevelCiPlugin extends AutoPlugin {
 
-  override def requires = GitHubActionsPlugin && GenerativePlugin && MimaPlugin
+  override def requires =
+    GitHubActionsPlugin && GenerativePlugin && MimaPlugin && TypelevelScalafixPlugin
+
   override def trigger = allRequirements
 
   object autoImport {
     def tlCrossRootProject: CrossRootProject = macro CrossRootProjectMacros.crossRootProjectImpl
+
+    val tlCheckHeaders =
+      settingKey[Boolean]("Check for presence of copyright headers in CI (default: true)")
+    val tlCheckFormatting = settingKey[Boolean](
+      "Check code is formatted according to the project's Scalafmt config in CI (default: true)")
+    val tlCheckMigrations = settingKey[Boolean](
+      "Check configured Scalafix migrations have been applied in CI (default: true)")
   }
 
+  import autoImport._
+
   override def buildSettings = Seq(
+    tlCheckHeaders := true,
+    tlCheckFormatting := true,
+    tlCheckMigrations := true,
     githubWorkflowPublishTargetBranches := Seq(),
-    githubWorkflowBuild := Seq(
-      WorkflowStep.Sbt(List("test"), name = Some("Test")),
-      WorkflowStep.Sbt(
-        List("mimaReportBinaryIssues"),
-        name = Some("Check binary compatibility"),
-        cond = Some(primaryJavaCond.value)
-      ),
-      WorkflowStep.Sbt(
-        List("doc"),
-        name = Some("Generate API documentation"),
-        cond = Some(primaryJavaCond.value)
+    githubWorkflowBuild := {
+      val checkHeadersCommands =
+        if (tlCheckHeaders.value) List("headerCheckAll") else Nil
+
+      val checkFormattingCommands =
+        if (tlCheckFormatting.value)
+          List("scalafmtCheckAll", "project /", "scalafmtSbtCheck")
+        else
+          Nil
+
+      val checkMigrationsStep =
+        if (tlCheckMigrations.value)
+          Seq(
+            WorkflowStep.Sbt(
+              List("scalafixAll --check"),
+              name = Some("Check Scalafix migrations"),
+              cond = Some(primaryJavaCond.value)
+            )
+          )
+        else Nil
+
+      val preBuildSteps = Seq(
+        WorkflowStep.Sbt(
+          checkHeadersCommands ++ checkFormattingCommands,
+          name = Some("Check headers and formatting"),
+          cond = Some(primaryJavaCond.value)
+        )
       )
-    ),
+
+      val buildStep = Seq(
+        WorkflowStep.Sbt(List("test"), name = Some("Test"))
+      )
+
+      val postBuildSteps = checkMigrationsStep ++ Seq(
+        WorkflowStep.Sbt(
+          List("mimaReportBinaryIssues"),
+          name = Some("Check binary compatibility"),
+          cond = Some(primaryJavaCond.value)
+        ),
+        WorkflowStep.Sbt(
+          List("doc"),
+          name = Some("Generate API documentation"),
+          cond = Some(primaryJavaCond.value)
+        )
+      )
+
+      Seq(
+        preBuildSteps,
+        buildStep,
+        postBuildSteps
+      ).flatten
+    },
     githubWorkflowJavaVersions := Seq(JavaSpec.temurin("8"))
   )
 
@@ -54,5 +108,4 @@ object TypelevelCiPlugin extends AutoPlugin {
     val java = githubWorkflowJavaVersions.value.head
     s"matrix.java == '${java.render}'"
   }
-
 }
```
I need to take a break now though and I seem to have lost the artifact upload steps, so I think I will wait to see what your changes look like instead 😆